File: src/desc.c
#include "desc.h"

#include <stdlib.h>
#include <stdio.h>
#include <stdbool.h>
#include <string.h>
#include <time.h>
#include <errno.h>
#include <err.h>

#include <archive.h>
#include <archive_entry.h>
#include <alpm_list.h>

#include "reader.h"
#include "package.h"
#include "util.h"

static inline int read_desc_list(struct archive_reader *reader, alpm_list_t **list)
{
    char *buf;
    int nbytes_r;
    while ((nbytes_r = archive_getline(reader, &buf)) > 0)
        *list = alpm_list_add(*list, buf);
    return nbytes_r;
}

static inline int read_desc_entry(struct archive_reader *reader, char **data)
{
    return archive_getline(reader, data);
}

static inline int read_desc_size(struct archive_reader *reader, size_t *data)
{
    _cleanup_free_ char *buf;
    int nbytes_r = archive_getline(reader, &buf);
    if (nbytes_r > 0)
        fromstr(buf, data);
    return nbytes_r;
}

static inline int read_desc_time(struct archive_reader *reader, time_t *data)
{
    _cleanup_free_ char *buf;
    int nbytes_r = archive_getline(reader, &buf);
    if (nbytes_r > 0)
        fromstr(buf, data);
    return nbytes_r;
}

void read_desc(struct archive *archive, struct pkg *pkg)
{
    struct archive_reader *reader = archive_reader_new(archive);

    for (;;) {
        char buf[8192];
        int nbytes_r = archive_fgets(reader, buf, sizeof(buf));
        if (nbytes_r < 0)
            break;

        if (streq(buf, "%FILENAME%")) {
            nbytes_r = read_desc_entry(reader, &pkg->filename);
        } else if (streq(buf, "%NAME%")) {
            _cleanup_free_ char *temp = NULL;
            nbytes_r = read_desc_entry(reader, &temp);
            if (!streq(temp, pkg->name))
                errx(EXIT_FAILURE, "database entry %%NAME%% and desc record are mismatched!");
        } else if (streq(buf, "%BASE%")) {
            nbytes_r = read_desc_entry(reader, &pkg->base);
        } else if (streq(buf, "%VERSION%")) {
            _cleanup_free_ char *temp = NULL;
            nbytes_r = read_desc_entry(reader, &temp);
            if (!streq(temp, pkg->version))
                errx(EXIT_FAILURE, "database entry %%VERSION%% and desc record are mismatched!");
        } else if (streq(buf, "%DESC%")) {
            nbytes_r = read_desc_entry(reader, &pkg->desc);
        } else if (streq(buf, "%GROUPS%")) {
            nbytes_r = read_desc_list(reader, &pkg->groups);
        } else if (streq(buf, "%CSIZE%")) {
            nbytes_r = read_desc_size(reader, &pkg->size);
        } else if (streq(buf, "%ISIZE%")) {
            nbytes_r = read_desc_size(reader, &pkg->isize);
        } else if (streq(buf, "%MD5SUM%")) {
            nbytes_r = read_desc_entry(reader, &pkg->md5sum);
        } else if (streq(buf, "%SHA256SUM%")) {
            nbytes_r = read_desc_entry(reader, &pkg->sha256sum);
        } else if(streq(buf, "%PGPSIG%")) {
            nbytes_r = read_desc_entry(reader, &pkg->base64sig);
        } else if (streq(buf, "%URL%")) {
            nbytes_r = read_desc_entry(reader, &pkg->url);
        } else if (streq(buf, "%LICENSE%")) {
            nbytes_r = read_desc_list(reader, &pkg->licenses);
        } else if (streq(buf, "%ARCH%")) {
            nbytes_r = read_desc_entry(reader, &pkg->arch);
        } else if (streq(buf, "%BUILDDATE%")) {
            nbytes_r = read_desc_time(reader, &pkg->builddate);
        } else if (streq(buf, "%PACKAGER%")) {
            nbytes_r = read_desc_entry(reader, &pkg->packager);
        } else if (streq(buf, "%REPLACES%")) {
            nbytes_r = read_desc_list(reader, &pkg->replaces);
        } else if (streq(buf, "%DEPENDS%")) {
            nbytes_r = read_desc_list(reader, &pkg->depends);
        } else if (streq(buf, "%CONFLICTS%")) {
            nbytes_r = read_desc_list(reader, &pkg->conflicts);
        } else if (streq(buf, "%PROVIDES%")) {
            nbytes_r = read_desc_list(reader, &pkg->provides);
        } else if (streq(buf, "%OPTDEPENDS%")) {
            nbytes_r = read_desc_list(reader, &pkg->optdepends);
        } else if (streq(buf, "%MAKEDEPENDS%")) {
            nbytes_r = read_desc_list(reader, &pkg->makedepends);
        } else if (streq(buf, "%CHECKDEPENDS%")) {
            nbytes_r = read_desc_list(reader, &pkg->checkdepends);
        } else if (streq(buf, "%FILES%")) {
            nbytes_r = read_desc_list(reader, &pkg->files);
        }

        if (nbytes_r < 0)
            break;
    }

    free(reader);
}

File: src/desc.rs
use std::fmt::Debug;
use std::ops::Deref;
use std::str::from_utf8;
use std::collections::hash_map;
use nom::{IResult, newline};
use package::{Package, Entry, Metadata};
use error::Result;

pub struct Parser {
    entry: Option<Entry>,
}

#[derive(Debug)]
enum Line<'a> {
    Entry(Entry),
    Line(&'a str),
    Seperator,
}


named!(header<Line>, do_parse!(
    entry: alt!(
        tag!("%FILENAME%")     => {|_| Line::Entry(Entry::Filename)}
      | tag!("%NAME%")         => {|_| Line::Entry(Entry::Name)}
      | tag!("%BASE%")         => {|_| Line::Entry(Entry::Base)}
      | tag!("%VERSION%")      => {|_| Line::Entry(Entry::Version)}
      | tag!("%DESC%")         => {|_| Line::Entry(Entry::Description)}
      | tag!("%GROUP%")        => {|_| Line::Entry(Entry::Groups)}
      | tag!("%CSIZE%")        => {|_| Line::Entry(Entry::PackageSize)}
      | tag!("%ISIZE%")        => {|_| Line::Entry(Entry::InstallSize)}
      | tag!("%SHA256SUM%")    => {|_| Line::Entry(Entry::SHA256Sum)}
      | tag!("%PGPSIG%")       => {|_| Line::Entry(Entry::PGPSig)}
      | tag!("%URL%")          => {|_| Line::Entry(Entry::Url)}
      | tag!("%LICENSE%")      => {|_| Line::Entry(Entry::License)}
      | tag!("%ARCH%")         => {|_| Line::Entry(Entry::Arch)}
      | tag!("%BUILDDATE%")    => {|_| Line::Entry(Entry::BuildDate)}
      | tag!("%PACKAGER%")     => {|_| Line::Entry(Entry::Packager)}
      | tag!("%REPLACES%")     => {|_| Line::Entry(Entry::Replaces)}
      | tag!("%DEPENDS%")      => {|_| Line::Entry(Entry::Depends)}
      | tag!("%CONFLICTS%")    => {|_| Line::Entry(Entry::Conflicts)}
      | tag!("%PROVIDES%")     => {|_| Line::Entry(Entry::Provides)}
      | tag!("%OPTDEPENDS%")   => {|_| Line::Entry(Entry::OptDepends)}
      | tag!("%MAKEDEPENDS%")  => {|_| Line::Entry(Entry::MakeDepends)}
      | tag!("%CHECKDEPENDS%") => {|_| Line::Entry(Entry::CheckDepends)}
      | tag!("%FILES%")        => {|_| Line::Entry(Entry::Files)}
      | tag!("%DELTAS%")       => {|_| Line::Entry(Entry::Deltas)}
    ) >>
    newline >>
    (entry)
));

named!(value<Line>, map_res!(
    take_until_and_consume!("\n"),
    |str| from_utf8(str).map(Line::Line)
));

named!(desc_entry<Line>, do_parse!(
    entry: alt!(do_parse!(newline >> (Line::Seperator)) | header | value) >>
    (entry)
));

pub trait DescCallbacks {
    fn on_name(&mut self, value: &str);
    fn on_version(&mut self, value: &str);
    fn on_arch(&mut self, value: &str);
    fn on_metadata(&mut self, entry: &Entry, value: &str);
}

impl DescCallbacks for Package {
    fn on_name(&mut self, value: &str) {
        assert_eq!(self.name, value);
    }

    fn on_version(&mut self, value: &str) {
        assert_eq!(self.version.deref(), value);
    }

    fn on_arch(&mut self, value: &str) {
        self.arch = value.into();
    }

    fn on_metadata(&mut self, entry: &Entry, value: &str) {
        let metadata = self.metadata.entry(entry.clone());
        match metadata {
            hash_map::Entry::Occupied(mut o) => {
                match *o.get_mut() {
                    Metadata::List(ref mut l) => l.push(value.into()),
                    _ => panic!("shouldn't happen but TODO"),
                };
            }
            hash_map::Entry::Vacant(v) => {
                v.insert((entry, value).into());
            }
        };
    }
}

impl Parser {
    pub fn new() -> Self {
        Parser { entry: None }
    }

    pub fn read<T>(&mut self, cbs: &mut T, b: &[u8]) -> Option<()>
        where T: DescCallbacks + Debug
    {
        let mut data = b;

        loop {
            if let IResult::Done(rest, result) = desc_entry(data) {
                match result {
                    Line::Entry(entry) => self.entry = Some(entry),
                    Line::Line(line) => {
                        match self.entry {
                            Some(Entry::Name) => cbs.on_name(line),
                            Some(Entry::Version) => cbs.on_version(line),
                            Some(Entry::Arch) => cbs.on_arch(line),
                            Some(ref entry) => cbs.on_metadata(&entry, line),
                            None => panic!("invalid state!"),
                        }
                    }
                    Line::Seperator => self.entry = None,
                }

                data = rest;
            } else {
                return None;
            }
        }
        Some(())
    }
}
